AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift

```diff
@@ -10,6 +10,177 @@ import Foundation
+                    .getOrCreateConnection(for: endpointConfig,
+                                              authService: authService,
+                                              authType: pluginOptions?.authType,
+                                              apiAuthProviderFactory: apiAuthProviderFactory)
+            } catch {
+                let error = APIError.operationError("Unable to get connection for api \(endpointConfig.name)", "", error)
+                fail(error)
+                return
+            }
+
+            // Create subscription
+
+            subscriptionItem = subscriptionConnection?.subscribe(requestString: request.document,
+                                                                 variables: request.variables,
+                                                                 eventHandler: { [weak self] event, _ in
+                self?.onAsyncSubscriptionEvent(event: event)
+            })
+        }
+    }
+    
+    // MARK: - Subscription callbacks
+    
+    private func onAsyncSubscriptionEvent(event: SubscriptionItemEvent) {
+        switch event {
+        case .connection(let subscriptionConnectionEvent):
+            onSubscriptionEvent(subscriptionConnectionEvent)
+        case .data(let data):
+            onGraphQLResponseData(data)
+        case .failed(let error):
+            onSubscriptionFailure(error)
+        }
+    }
+
+    private func onSubscriptionEvent(_ subscriptionConnectionEvent: SubscriptionConnectionEvent) {
+        switch subscriptionConnectionEvent {
+        case .connecting:
+            let subscriptionEvent = GraphQLSubscriptionEvent<R>.connection(.connecting)
+            send(subscriptionEvent)
+        case .connected:
+            let subscriptionEvent = GraphQLSubscriptionEvent<R>.connection(.connected)
+            send(subscriptionEvent)
+        case .disconnected:
+            let subscriptionEvent = GraphQLSubscriptionEvent<R>.connection(.disconnected)
+            send(subscriptionEvent)
+            finish()
+        }
+    }
+
+    private func onSubscriptionConnectionState(_ subscriptionConnectionState: SubscriptionConnectionState) {
+        let subscriptionEvent = GraphQLSubscriptionEvent<R>.connection(subscriptionConnectionState)
+        send(subscriptionEvent)
+
+        if case .disconnected = subscriptionConnectionState {
+            finish()
+        }
+    }
+
+    private func onGraphQLResponseData(_ graphQLResponseData: Data) {
+        do {
+            let graphQLResponseDecoder = GraphQLResponseDecoder(request: request, response: graphQLResponseData)
+            let graphQLResponse = try graphQLResponseDecoder.decodeToGraphQLResponse()
+            send(.data(graphQLResponse))
+        } catch let error as APIError {
+            fail(error)
+        } catch {
+            // TODO: Verify with the team that terminating a subscription after failing to decode/cast one
```

Comment:
Please add a subtask to create a unit test for this scenario.

Reply:
Added subtask for internal tracking